packages/database/src/migrations/20210511042757-MoveLESMISVisualsToNewDashboards-modifies-data.js

```diff
@@ -0,0 +1,181 @@
+      concat("organisationUnitCode", '_', REPLACE(name, ' ', '_'), '_', REPLACE((array_agg(distinct "userGroup"))[1], ' ', '_')) as code
+    FROM "dashboardGroup"
+    WHERE 'laos_schools' = ANY("projectCodes")
+    GROUP BY name, "organisationUnitCode";
+  `);
+  await deleteArrayCatAgg(db);
+
+  // Array types need to be formated specifically for insertObject
+  for (const currentDashboard of dashboards.rows) {
+    await insertObject(db, 'dashboard', {
+      id: generateId(),
+      ...currentDashboard,
+      entity_types: `{${arrayToDoubleQuotedDbString(currentDashboard.entity_types)}}`,
+      project_codes: `{${arrayToDoubleQuotedDbString(currentDashboard.project_codes)}}`,
+    });
+  }
+};
+
+const createLegacyDashboardItems = async db => {
+  const dashboardReports = await db.runSql(`
+    SELECT * from "dashboardReport"
+    WHERE id IN
+      (SELECT unnest("dashboardReports") from "dashboardGroup"
+       WHERE 'laos_schools' = ANY("projectCodes"));
+  `);
+
+  for (const currentReport of dashboardReports.rows) {
+    const { id: code, viewJson: config, ...legacyObject } = currentReport;
+    const legacyId = generateId();
+    const permissionGroups = await db.runSql(`
+      SELECT array_agg(DISTINCT "userGroup") as permission_groups
+      FROM "dashboardGroup"
+      WHERE '${code}' = ANY("dashboardReports");
+    `);
+    await db.runSql(`
+      -- Drop trigger because this change notification can blow out the size limit on pg_notify
+      DROP TRIGGER IF EXISTS legacy_report_trigger
+      ON legacy_report;
+
+      INSERT INTO legacy_report
+      VALUES ('${legacyId}',
+              '${code}',
+              '${currentReport.dataBuilder}',
+              '${JSON.stringify(currentReport.dataBuilderConfig)}',
+              '${JSON.stringify(currentReport.dataServices)}'
+      );
+
+      -- Recreate triggers
+      CREATE TRIGGER legacy_report_trigger
+        AFTER INSERT OR UPDATE or DELETE
+        ON legacy_report
+        FOR EACH ROW EXECUTE PROCEDURE notification();
+    `);
+    await db.runSql(`
+      INSERT INTO dashboard_item
+      VALUES (
+        '${generateId()}',
+        '${code}',
+        '${JSON.stringify(config).replace(/'/g, "''")}',
+        '{${permissionGroups.rows[0].permission_groups}}',
+        '${code}',
+         ${true}
+      )
+    `);
+  }
+};
+
+const createDashboardRelationsFromDashboardGroups = async db => {
+  const dashboardRelations = await db.runSql(`
+    SELECT
+      dashboard.id AS dashboard_id,
+      unnest("dashboardReports") AS child_id
+    FROM "dashboardGroup"
+    INNER JOIN dashboard
+      ON dashboard.name = "dashboardGroup".name
+    WHERE 'laos_schools' = ANY("projectCodes")
+    GROUP BY dashboard.id, child_id;
+  `);
+
+  for (const currentDashboardRelation of dashboardRelations.rows) {
+    const childId = await db.runSql(`
+      SELECT id FROM dashboard_item
+      WHERE code = '${currentDashboardRelation.child_id}'
+    `);
```

Comment:
```suggestion
    const { id: childId } = (await db.runSql(`
      SELECT id FROM dashboard_item
      WHERE code = '${currentDashboardRelation.child_id}'
    `))[0];
```
This is the sort of thing I'd be happy to pull out into a migration util called `findOne`, which can be imported (because it doesn't make any assumptions about our database structure, so shouldn't change over time and affect past migrations)

Reply:
Oh and my reasoning here is I prefer variables to be named with no surprises, and I was surprised that something called `childId` held a structure like `[{ id: 'xxx', ... }]`